[AUTH-1481] [AUTH-1492] Make tests compatible with Remote selenium; DRY some things out for maintainability

tomthorogood opened the pull request:
# Change Summary

When reading the actual code changes, you may want to **go commit-by-commit, and _skip_ b82447c** (the 'webdriver-recorder 4.0' commit). That commit is huge, but mostly repeated changes, which are explained in detail below (to save you from having to read the monstrosity to figure it out). 

[Storyboards for this change (manually uploaded)](https://identity-artifact.iamdev.s.uw.edu/idp-web-tests/sha/b82447c.html)

---

## Why the big diff?

Most of the changes can be summed up with the following examples:

### Example 1: Logging in to the idp

This:

```python
#  Happy case
browser.get(f'{sp_url(sp)}/shib{self.test_env}mfa')
browser.wait_for_tag('p', 'Please sign in')
browser.send_inputs(netid3, password)
browser.click(Locators.submit_button)
browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
```

was reduced to this:

```python
browser.get(sp_shib_url(sp, append='mfa'))
log_in_netid(browser, netid3, match_service_provider=sp)
```
### Example 2: Entering the duo passcode

Before:

```python
browser.execute_script("document.getElementsByClassName('passcode-input')[0].value=arguments[0]", passcode)
browser.snap()
browser.execute_script("document.getElementById('passcode').click()")
browser.wait_for_tag('span', 'Incorrect passcode. Enter a passcode from Duo Mobile.')
browser.execute_script("document.getElementById('passcode').click()")
browser.execute_script("document.getElementsByClassName('passcode-input')[0].value=arguments[0]", self.passcode)
browser.snap()
browser.execute_script("document.getElementById('passcode').click()")
browser.switch_to.default_content()
browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
```

After:

```python
enter_duo_passcode(browser, match_service_provider=sp)   # Happy case
```

### 

Since those two flows are repeated dozens of times each, you can see how factoring them resulted in such a large, but honestly pretty boring, diff. The examples only show happy cases; the refactor accounts for all variations on this. Keep reading for more information and links to the _important_ bits of code.

---

# Developer Impact

## AWS Credentials

AWS Credentials are no longer required setup for developers. Instead, those credentials will be 
accessed using the GOOGLE_APPLICATION_CREDENTIALS that are already required for 
other packages.

## Setting up and running tests

In order to prepare to use the shared poetry docker image,
I migrated this package to use poetry, which simplifies the setup process 
and makes it easier for everyone, including docker images, to set up and 
run these tests.

After this change, developers should follow the steps in README.md to ensure they are properly set up. 


## Fixtures

### Browser fixtures were overhauled

The webdriver-recorder 4.0 changes rendered some of the fixtures needed here obsolete, but this test suite also has several edge cases with regards to cookies that can only be handled by creating what Glen coined "fresh" instances. To ensure these fresh instances are created safely with respect to their current environment, I added two fixtures that handle all the complexity of setting them up:

- `fresh_browser` will configure a browser instance, connect it to a running chromedriver service, and clean up after your test is finished.
- `fresh_class_browser` can be applied to class-based tests to automatically set `self.browser` to a "fresh" instance whose lifetime is scoped to the class; as with `fresh_browser`, this instance cleans up after itself when all tests in the class have finished.


### Updates the `login_submit_form` fixture:

- Renamed this fixture to `log_in_netid` so that the flow is descriptive of what the browser is doing, rather than describing a DOM element. 
- The `password` argument is now optional, and will default to the secret (i.e., correct) password.
- Automatically detects whether the entered password is the correct one, and asserts the success case if so. 
- Use `assert_success=False` if you know that the login won't succeed
- Use `select_iframe=False` if you know the iframe context is already selected
- Use `match_service_provider=sp` when you expect the login to succeed (otherwise, only the success message, not the idp domain, will be matched).

### Updates the `two_fa_submit_form` fixture:

- Renamed this fixture to `enter_duo_passcode` so that the flow is descriptive of what the browser is doing, and also to remove the naming issue introduced by the number 2.
- The `passcode` argument is now optional, and will default to the secret (i.e., correct) passcode.
- Automatically detects whether the entered passcode is the correct one, and asserts either the success state or the invalid passcode state. You can toggle both checks using `assert_success=True|False` and `assert_failure=True|False`, respectively.
- Use `match_service_provider=sp` when you expect the login to succeed (otherwise, only the success message, not the idp domain, will be matched).

### Adds the fixture `sp_shib_url`

Since almost every test case used the pattern `f"{sp_url(sp)/shib{test_env}"`, I refactored this into a fixture and replaced all instances.

Use it like so:

```python
# assuming test_env == 'prod' and sp == ServiceProviderInstance.diafine6

sp_shib_url(sp)   # https://diafine6.sandbox.iam.s.uw.edu/shibprod
sp_shib_url(sp, append='mfa')  # https://diafine6.sandbox.iam.s.uw.edu/shibprodmfa
```

This accounts for the current `test_env`.

Commit:
Migrate to poetry and update developer setup instructions.

Commit:
Auto-fetch AWS credentials from secrets; update documentation.

Also:

- Fixes a bug that would cause TestOptions fields to conflict with arguments added by other pytest plugins.

Commit:
Fix the broken `sp_manager.py`

Commit:
- Add some helpful log messages to the test SP operations
- Properly scopes several fixtures to the session

Commit:
Add reusable locator references for the duo iframe and duo passcode button

Commit:
Migrate to webdriver-recorder 4.0 and refactor for maintainability

- Drops several fixtures that are no longer needed after the webdriver-recorder 4.0 update
- Use class fixtures properly (I learned something new!)
- Properly configure and scope "fresh" instances.
- Add fixture: `sp_shib_url`, which optionally accepts any string to concatenate to the url.
- Add fixture: `enter_duo_passcode` to DRY out some complex logic. See docstrings. Automatically runs all assertions unless directed not to.
- Rename `login_submit_form` to `log_in_netid` so that it's verby, runs assertions as part of the fixture, unless directed not to.
- Adds a `fresh_browser` fixture which handles all logic and configuration for function-scoped "fresh" browsers
- Adds a `fresh_class_browser` fixture which handles all logic and configuration for class-scoped "fresh" browsers.
- Speeds up testing by sharing a single chromedriver connection when running locally for all "fresh" browsers.
- Ensures that secret values are treated as secrets, and not printed in debug output (without intent).

tomthorogood commented:
I'll handle this merge conflict and update the PR sometime today

tomthorogood commented:
Updated to fix merge conflicts

goulter commented on tests/access_control/test_auto_2fa.py:
```diff
@@ -1,118 +1,95 @@
-            fresh_browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
+            self.browser.get(self.sp_shib_url(sp))
+            self.enter_duo_passcode(self.browser)
+            self.browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
```
Looks like line 31 would be redundant based on the definition of `self.enter_duo_passcode`

tomthorogood replied:
Correct, thanks!

goulter commented on tests/access_control/test_auto_2fa.py:
```diff
@@ -1,118 +1,95 @@
-            self.login_submit_form(fresh_browser, netid3, self.password)
-            fresh_browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
+            with self.browser.tab_context():
+                self.browser.get(self.sp_shib_url(sp))
```
what is `self.browser.tab_context():`?

tomthorogood replied:
Simply opens a tab, then closes it at the end of the block. Only used here because you had previously explicitly opened and closed a tab in this particular test.

https://github.com/UWIT-IAM/webdriver-recorder/blob/master/webdriver_recorder/browser.py#L264-L270

tomthorogood replied:
(Not sure if the tab is actually necessary, but preserved the flow you wrote)

goulter commented on tests/access_control/test_auto_2fa.py:
```diff
@@ -1,118 +1,95 @@
-        AC-1 Part D. No prompts on diafine7.
+        AC-1 Part D
         """
-        fresh_browser = Chrome()
```
What is the replacement for the new browser instance that was getting created on line 73?

tomthorogood replied:
`self.browser` is always a `fresh_browser` for every test in the `access_control` package, if you look at the `AccessControlTestBase` class.

goulter replied:
I see it now, cool.

goulter commented on tests/test_attributes.py:
```diff
@@ -57,36 +54,17 @@ def test_attributes(browser, netid, utils, sp_url, sp_domain, secrets, test_env)
-                # mail
-                assert f'{key} = {value}' in attribute_data
-
-        fresh_browser.close()
```
Lines 73-78 and 80-86 should not be removed, they account for testing `eduPersonScopedAffiliation`.

tomthorogood replied:
I feel like this might have been a merge conflict problem. I'll restore the intent here. I wouldn't have purposely deleted that stuff.

tomthorogood replied:
I think these weren't so much "removed" as "accidentally merge-conflicted away" --- i'll make sure to add those test cases back

goulter replied:
Thx

goulter commented on tests/test_2fa_duo.py:
```diff
@@ -3,264 +3,205 @@
-    # chrome_options.add_experimental_option("prefs", {"profile.default_content_settings.cookies": 1})
-    # fresh_browser = Chrome(options=chrome_options)
-    fresh_browser = Chrome()
+    fresh_browser
```
Does `fresh_browser` have a value?

tomthorogood replied:
of course it has a value, it's a fixture, but no, this line shouldn't be here.

goulter reviewed (commented):
A few questions, one of which may spawn a change request :)

Commit:
- Remove `session_browser` as a dependency for building a `fresh_browser`

- Remove unused function (`add_suffix`)

- Reinstate accidentally removed test case from `test_attributes.py`

- Remove redundant statement, unused statement from `test_2fa_duo.py`
